File: infrastructure/adapters/in/DoctorController.js
// Controlador REST para doctores
const express = require('express');

function createDoctorRouter(doctorService) {
    const router = express.Router();

    // POST - crear un doctor
    router.post('/', (req, res) => {
        const { nombre, especialidad } = req.body;
        if (!nombre || !especialidad) {
            return res.status(400).json({ error: 'nombre y especialidad son obligatorios' });
        }
        const doctor = doctorService.createDoctor(nombre, especialidad);
        res.status(201).json(doctor);
    });

    // GET - consultar todos los doctores
    router.get('/', (req, res) => {
        const doctors = doctorService.getAllDoctors();
        res.json(doctors);
    });

    // GET - consultar doctor por id
    router.get('/:id', (req, res) => {
        const doctor = doctorService.getDoctorById(Number(req.params.id));
        if (!doctor) {
            return res.status(404).json({ error: 'Doctor no encontrado' });
        }
        res.json(doctor);
    });

    // DELETE - eliminar un doctor
    router.delete('/:id', (req, res) => {
        const deleted = doctorService.deleteDoctor(Number(req.params.id));
        if (!deleted) {
            return res.status(404).json({ error: 'Doctor no encontrado' });
        }
        res.json({ message: 'Doctor eliminado correctamente' });
    });

    return router;
}

module.exports = createDoctorRouter;

File: infrastructure/adapters/in/doctor_controller.py
# Controlador REST para doctores
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

class DoctorRequest(BaseModel):
    nombre: str
    especialidad: str

def create_doctor_router(doctor_service):
    router = APIRouter()

    # POST - para crear un doctor
    @router.post("/doctors", status_code=201)
    def create_doctor(data: DoctorRequest):
        doctor = doctor_service.create_doctor(data.nombre, data.especialidad)
        return doctor.to_dict()

    # GET - para consultar todos los doctores
    @router.get("/doctors")
    def get_all_doctors():
        doctors = doctor_service.get_all_doctors()
        return [d.to_dict() for d in doctors]

    # GET - consultar doctor por id
    @router.get("/doctors/{id}")
    def get_doctor_by_id(id: int):
        doctor = doctor_service.get_doctor_by_id(id)
        if not doctor:
            raise HTTPException(status_code=404, detail="Doctor no encontrado")
        return doctor.to_dict()

    # DELETE - para eliminar un doctor
    @router.delete("/doctors/{id}")
    def delete_doctor(id: int):
        deleted = doctor_service.delete_doctor(id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Doctor no encontrado")
        return {"message": "Doctor eliminado correctamente"}

    return router
